File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.java
/*
 * Copyright 2016, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.android.architecture.blueprints.todoapp.taskdetail;

import android.content.Context;

import androidx.annotation.Nullable;

import com.example.android.architecture.blueprints.todoapp.TaskViewModel;
import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepository;
import com.example.android.architecture.blueprints.todoapp.tasks.TasksFragment;


/**
 * Listens to user actions from the list item in ({@link TasksFragment}) and redirects them to the
 * Fragment's actions listener.
 */
public class TaskDetailViewModel extends TaskViewModel {

    @Nullable
    private TaskDetailNavigator mTaskDetailNavigator;

    public TaskDetailViewModel(Context context, TasksRepository tasksRepository) {
        super(context, tasksRepository);
    }

    public void setNavigator(TaskDetailNavigator taskDetailNavigator) {
        mTaskDetailNavigator = taskDetailNavigator;
    }

    public void onActivityDestroyed() {
        // Clear references to avoid potential memory leaks.
        mTaskDetailNavigator = null;
    }

    /**
     * Can be called by the Data Binding Library or the delete menu item.
     */
    public void deleteTask() {
        super.deleteTask();
        if (mTaskDetailNavigator != null) {
            mTaskDetailNavigator.onTaskDeleted();
        }
    }

    public void startEditTask() {
        if (mTaskDetailNavigator != null) {
            mTaskDetailNavigator.onStartEditTask();
        }
    }
}

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt
package com.example.android.architecture.blueprints.todoapp.taskdetail

import android.app.Application
import androidx.annotation.StringRes
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import com.example.android.architecture.blueprints.todoapp.Event
import com.example.android.architecture.blueprints.todoapp.R
import com.example.android.architecture.blueprints.todoapp.data.Task
import com.example.android.architecture.blueprints.todoapp.data.source.TasksDataSource
import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepository

/**
 * Listens to user actions from the list item in ([TasksFragment]) and redirects them to the
 * Fragment's actions listener.
 */
open class TaskDetailViewModel(context: Application, private val tasksRepository: TasksRepository) :
    AndroidViewModel(context), TasksDataSource.GetTaskCallback {

    private val _task = MutableLiveData<Task>()

    private val _isDataAvailable = MutableLiveData<Boolean>()

    private val _dataLoading = MutableLiveData<Boolean>()

    val editTaskCommand = MutableLiveData<Event<Any>>()

    val deleteTaskCommand = MutableLiveData<Event<Any>>()

    private val _snackbarText = MutableLiveData<Event<Int>>()

    // This LiveData depends on another so we can use a transformation.
    val completed: LiveData<Boolean> = Transformations.map(_task) {
            input -> input.isCompleted
    }

    val snackbarMessage: LiveData<Event<Int>>
        get() = _snackbarText

    val task: LiveData<Task>
        get() = _task

    val isDataAvailable: LiveData<Boolean>
        get() = _isDataAvailable

    val dataLoading: LiveData<Boolean>
        get() = _dataLoading

    protected val taskId: String?
        get() = _task.value!!.id

    fun deleteTask() {
        if (_task.value != null) {
            tasksRepository.deleteTask(_task.value!!.id)
            deleteTaskCommand.value = Event(Any())
        }
    }

    fun editTask() {
        editTaskCommand.value = Event(Any())
    }

    fun setCompleted(completed: Boolean) {
        if (_dataLoading.value!!) {
            return
        }
        val task = this._task.value
        if (completed) {
            tasksRepository.completeTask(task!!)
            showSnackbarMessage(R.string.task_marked_complete)
        } else {
            tasksRepository.activateTask(task!!)
            showSnackbarMessage(R.string.task_marked_active)
        }
    }

    fun start(taskId: String?) {
        if (taskId != null) {
            _dataLoading.value = true
            tasksRepository.getTask(taskId, this)
        }
    }

    private fun setTask(task: Task?) {
        this._task.value = task
        _isDataAvailable.value = task != null
    }

    override fun onTaskLoaded(task: Task) {
        setTask(task)
        _dataLoading.value = false
    }

    override fun onDataNotAvailable() {
        _task.value = null
        _dataLoading.value = false
        _isDataAvailable.value = false
    }

    fun onRefresh() {
        if (_task.value != null) {
            start(_task.value!!.id)
        }
    }

    private fun showSnackbarMessage(@StringRes message: Int) {
        _snackbarText.value = Event(message)
    }
}